lib/course_planner/tasks.ex

```diff
@@ -47,59 +23,81 @@ defmodule CoursePlanner.Tasks do
     end
   end
 
-  def get_unassigned(sort_opt, now) do
+  def get_availables(sort_opt, id, now) do
     sort_opt
     |> task_query()
-    |> where([t], is_nil(t.user_id))
     |> where([t], t.finish_time > ^now)
     |> Repo.all()
+    |> Enum.reject(fn(task) ->
+           length(task.volunteers) >= task.max_volunteers or
+           Enum.any?(task.volunteers, &(&1.id == id))
+       end)
   end
 
   def get_past(sort_opt, id, now) do
     sort_opt
     |> task_query()
-    |> where([t], t.user_id == ^id)
     |> where([t], t.finish_time < ^now)
     |> Repo.all()
+    |> Enum.reject(fn(task) ->
+         not Enum.any?(task.volunteers, &(&1.id == id))
+       end)
   end
 
   def get_for_user(sort_opt, id, now) do
     sort_opt
     |> task_query()
-    |> where([t], t.user_id == ^id)
     |> where([t], t.finish_time > ^now)
     |> Repo.all()
+    |> Enum.reject(fn(task) ->
+         not Enum.any?(task.volunteers, &(&1.id == id))
+       end)
   end
 
-  def task_query(sort_opt), do: Task |> sort(sort_opt) |> preload(:user)
+  def task_query(sort_opt) do
+     Task
+     |> sort(sort_opt)
+     |> join(:left, [t], v in assoc(t, :volunteers))
+     |> preload([t, v], [volunteers: v])
+  end
 
   defp sort(query, nil), do: query
   defp sort(query, "fresh"), do: order_by(query, [t], desc: t.updated_at)
   defp sort(query, "closest"), do: order_by(query, [t], asc: t.finish_time)
 
-  def grab(task_id, user_id, now) do
-    with {:ok, task}              <- get(task_id),
-      %{valid?: true} = changeset <- Task.changeset(task),
-      {:ok, changeset}            <- validate_finish_time(changeset, now),
-      {:ok, changeset}            <- validate_already_assigned(changeset, task)
-    do
+  def grab(task_id, volunteer_id) do
+    with {:ok, task} <- get(task_id),
+        %{valid?: true} = changeset <- Task.changeset(task)
+     do
```

Comment:
I added the Enum.uniq and a test for this. but in any case even without the enum.uniq it was not adding a duplicated relation.